README.md
# cordova-plugin-screensize

![Platform](https://img.shields.io/badge/platform-android%20%7C%20ios%20%7C%20osx%20%7C%20browser-lightgrey.svg)
[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=R7STJ6V2PNEMA)

A cordova plugin to get the size of the device screen.

## Installation

```
cordova plugin add cordova-plugin-screensize
```

## Supported Platforms

- Android
- iOS
- OSX
- Browser

## Usage

```js
document.addEventListener("deviceready", onDeviceReady, false);

function onDeviceReady() {
  window.plugins.screensize.get(successCallback, errorCallback);
}

function successCallback(result) {
  console.log(result);
}
```

The plugin returns a JSON object with the screen sizes.

## Android Platform Quirks

Return values:

* `width` &lt;Number&gt; − screen width in pixels
* `height` &lt;Number&gt; − screen height in pixels
* `xdpi` &lt;Number&gt; − physical pixels per inch of the screen in the X dimension
* `ydpi` &lt;Number&gt; − physical pixels per inch of the screen in the Y dimension
* `diameter` &lt;Number&gt; − screen diameter in inches (rounded to two decimals)
* `densityValue` &lt;Number&gt; − density point (`0.75`, `1.0`, `1.5`, `2.0`, `3.0`, `4.0`)
* `densityBucket` &lt;String&gt; − density bucket (`ldpi`, `mdpi`, `hdpi`, `xhdpi`, `xxhdpi`, `xxxhdpi`)

About density properties see this [Android docs](https://developer.android.com/guide/practices/screens_support.html) and this [Device metrics](https://design.google.com/devices/) page.

### Known Issues

The correct device metrics setting is the manufacturer's responsibility. In a few cases the `xdpi` and `ydpi` values are wrong and the diameter calculation returns an incorrect number.

Devices with wrong settings (known so far):

- BLU Life XL (model `BLU LIFE XL`, calculated diameter 8.62" vs. real diameter 5.5")
- Samsung Galaxy SIII CDMA (model `SPH-L710`, calculated diameter 9.17" vs. real diameter 4.8")
- ZTE nubia Z7 Max (model `NX505J`, calculated diameter 13.77" vs. real diameter 5.5")

## iOS Platform Quirks

iOS has no public API for getting the device's PPI.

The plugin returns the [rendered pixels](http://www.paintcodeapp.com/news/ultimate-guide-to-iphone-resolutions) of the screen.

Return values:

* `width` &lt;Number&gt; − rendered pixel width
* `height` &lt;Number&gt; − rendered pixel height
* `scale` &lt;Number&gt; − render scale (1× or 2× or 3×)

## OSX Platform Quirks

Return values:

* `width` &lt;Number&gt; − screen width in pixel
* `height` &lt;Number&gt; − screen height in pixel
* `physicalWidth` &lt;Number&gt; − width of the display in millimeters
* `physicalHeight` &lt;Number&gt; − height of the display in millimeters
* `dpi` &lt;Number&gt; − physical pixels per inch of the display

## Browser Platform Quirks

The plugin returns the dimensions of the viewport. You can't get the real physical dimensions or the actual DPI of the browser ( [see](http://stackoverflow.com/a/21767407) ).

Return values:

* `width` &lt;Number&gt; − viewport width in pixels
* `height` &lt;Number&gt; − viewport height in pixels

## Author

#### Peter Bakondy

- https://github.com/pbakondy


## LICENSE

**cordova-plugin-screensize** is licensed under the MIT Open Source license. For more information, see the LICENSE file in this repository.


CHANGELOG.md
# Changelog

## v1.3.1

- add densityValue and densityBucket to Android platform

## v1.3.0

- add OSX platform

## v1.2.0

- add `xdpi` and `ydpi` values to Android platform
- add "Known Issues" section to README#Android

## v1.1.0

- add iOS and Browser platforms


tests/tests.js
exports.defineAutoTests = function() {
  describe('Screen Size Information (window.plugins.screensize)', function () {
    it('should exist', function() {
      expect(window.plugins.screensize).toBeDefined();
    });

    it('should contain get that is a function', function() {
      expect(window.plugins.screensize.get).toBeDefined();
      expect(typeof window.plugins.screensize.get).toBe('function');
    });

  });
};

exports.defineManualTests = function(contentEl, createActionButton) {
  var logMessage = function (message, color) {
    var log = document.getElementById('info');
    var logLine = document.createElement('div');
    if (color) {
      logLine.style.color = color;
    }
    logLine.innerHTML = message;
    log.appendChild(logLine);
  };

  var clearLog = function () {
    var log = document.getElementById('info');
    log.innerHTML = '';
  };

  var device_tests = '<h3>Press Dump Screensize button to get SIM information</h3>' +
    '<div id="dump_screensize"></div>' +
    'Expected result: Status box will get updated with Screensize info.';

  contentEl.innerHTML = '<div id="info"></div>' + device_tests;

  createActionButton('Dump Screensize', function() {
    clearLog();
    window.plugins.screensize.get(
      function(result) {
        logMessage(JSON.stringify(result));
      },
      function(error) {
        logMessage(JSON.stringify(error));
      }
    );
  }, 'dump_screensize');
};


src/osx/ScreenSize.m
// https://developer.apple.com/library/mac/documentation/Cocoa/Reference/ApplicationKit/Classes/NSScreen_Class/
// https://developer.apple.com/library/prerelease/content/documentation/GraphicsImaging/Conceptual/QuartzDisplayServicesConceptual/Articles/DisplayInfo.html

#import "ScreenSize.h"

@implementation ScreenSize

- (NSDictionary*)screenProperties
{

    NSScreen *screen = [NSScreen mainScreen];
    NSDictionary *description = [screen deviceDescription];
    NSSize displayPixelSize = [[description objectForKey:NSDeviceSize] sizeValue];
    CGSize displayPhysicalSize = CGDisplayScreenSize(
                [[description objectForKey:@"NSScreenNumber"] unsignedIntValue]);

    return @{
             @"width": [NSNumber numberWithFloat:displayPixelSize.width],
             @"height": [NSNumber numberWithFloat:displayPixelSize.height],
             @"physicalWidth": [NSNumber numberWithFloat:displayPhysicalSize.width],
             @"physicalHeight": [NSNumber numberWithFloat:displayPhysicalSize.height],
             @"dpi": [NSNumber numberWithFloat:(displayPixelSize.width / displayPhysicalSize.width) * 25.4f]
    };
}

- (void)get:(CDVInvokedUrlCommand*)command
{
    NSDictionary* screenProperties = [self screenProperties];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:screenProperties];

    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

@end


src/android/com/pbakondy/Screensize.java
package com.pbakondy;

import android.app.Activity;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;

import org.json.JSONObject;
import org.json.JSONArray;
import org.json.JSONException;

import android.content.Context;
import android.util.DisplayMetrics;

public class Screensize extends CordovaPlugin {

  @Override
  public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
    if (action.equals("get")) {

      // http://developer.android.com/reference/android/util/DisplayMetrics.html

      DisplayMetrics dm = new DisplayMetrics();
      this.cordova.getActivity().getWindowManager().getDefaultDisplay().getMetrics(dm);

      int width = dm.widthPixels;
      int height = dm.heightPixels;

      double wi = (double)width / (double)dm.xdpi;
      double he = (double)height / (double)dm.ydpi;
      double x = Math.pow(wi, 2);
      double y = Math.pow(he, 2);

      double screenInches = Math.round(Math.sqrt(x + y) * 100.0) / 100.0;

      JSONObject result = new JSONObject();

      try {
        result.put("width", width);
        result.put("height", height);
        result.put("diameter", screenInches);
      } catch (JSONException e) {}

      callbackContext.success(result);

      return true;
    } else {
      return false;
    }
  }
}


src/android/com/pbakondy/ScreenSize.java
package com.pbakondy;

import android.app.Activity;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;

import org.json.JSONObject;
import org.json.JSONArray;
import org.json.JSONException;

import android.content.Context;
import android.util.DisplayMetrics;

public class ScreenSize extends CordovaPlugin {

  @Override
  public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
    if (action.equals("get")) {

      // http://developer.android.com/reference/android/util/DisplayMetrics.html

      DisplayMetrics dm = new DisplayMetrics();
      this.cordova.getActivity().getWindowManager().getDefaultDisplay().getMetrics(dm);

      int width = dm.widthPixels;
      int height = dm.heightPixels;

      double xdpi = (double)dm.xdpi;
      double ydpi = (double)dm.ydpi;
      double wi = (double)width / xdpi;
      double he = (double)height / ydpi;
      double x = Math.pow(wi, 2);
      double y = Math.pow(he, 2);

      double screenInches = Math.round(Math.sqrt(x + y) * 100.0) / 100.0;

      double density_value = (double)dm.density;
      String density_bucket = "";

      // https://design.google.com/devices/
      // https://developer.android.com/guide/practices/screens_support.html

      if (density_value < 1.0) {
        density_bucket = "ldpi";
      } else if (density_value < 1.5) {
        density_bucket = "mdpi";
      } else if (density_value < 2.0) {
        density_bucket = "hdpi";
      } else if (density_value < 3.0) {
        density_bucket = "xhdpi";
      } else if (density_value < 4.0) {
        density_bucket = "xxhdpi";
      } else {
        density_bucket = "xxxhdpi";
      }

      JSONObject result = new JSONObject();

      try {
        result.put("width", width);
        result.put("height", height);
        result.put("diameter", screenInches);
        result.put("xdpi", xdpi);
        result.put("ydpi", ydpi);
        result.put("densityValue", density_value);
        result.put("densityBucket", density_bucket);
      } catch (JSONException e) {}

      callbackContext.success(result);

      return true;
    } else {
      return false;
    }
  }
}
